sycl/include/sycl/detail/accessor_iterator.hpp

```diff
@@ -0,0 +1,439 @@
+      __decrement();
+    else
+      __increment();
+    return *this;
+  }
+
+  __accessor_iterator_base operator++(int) {
+    auto _Old = *this;
+    ++(*this);
+    return _Old;
+  }
+
+  __accessor_iterator_base &operator--() {
+    if constexpr (_IsReverse)
+      __increment();
+    else
+      __decrement();
+    return *this;
+  }
+
+  __accessor_iterator_base operator--(int) {
+    auto _Old = *this;
+    --(*this);
+    return _Old;
+  }
+
+  __accessor_iterator_base &operator+=(difference_type _N) {
+    // iterator && N > 0 -> forward
+    // iterator && N < 0 -> backwards
+    // reverse iterator && N > 0 -> backwards
+    // reverse iterator && N < 0 -> forward
+    bool _BackwardsDirection = !_IsReverse ^ (_N > 0);
+    if (_BackwardsDirection)
+      __adjustBackwards(_N);
+    else
+      __adjustForward(_N);
+    return *this;
+  }
+
+  __accessor_iterator_base &operator-=(difference_type _N) {
+    // iterator && N > 0 -> backwards
+    // iterator && N < 0 -> forward
+    // reverse iterator && N > 0 -> forward
+    // reverse iterator && N < 0 -> backwards
+    bool _ForwardDirection = !_IsReverse ^ (_N > 0);
+    if (_ForwardDirection)
+      __adjustForward(_N);
+    else
+      __adjustBackwards(_N);
+    return *this;
+  }
+
+  difference_type operator-(const __accessor_iterator_base &_Rhs) {
+    if (_Rhs._MLinearCurrent > _MLinearCurrent)
+      return _Rhs._MLinearCurrent - _MLinearCurrent;
+    else
+      return _MLinearCurrent - _Rhs._MLinearCurrent;
+  }
+
+  bool operator<(const __accessor_iterator_base<_Dimensions> &_Other) const {
+    return _MLinearCurrent < _Other._MLinearCurrent;
+  }
+
+  bool operator>(const __accessor_iterator_base<_Dimensions> &_Other) const {
+    return _Other < *this;
+  }
+
+  bool operator<=(const __accessor_iterator_base<_Dimensions> &_Other) const {
+    return !(*this > _Other);
+  }
+
+  bool operator>=(const __accessor_iterator_base<_Dimensions> &_Other) const {
+    return !(*this < _Other);
+  }
+
+  bool operator==(const __accessor_iterator_base<_Dimensions> &_Other) const {
+    return _MLinearCurrent == _Other._MLinearCurrent;
+  }
+
+  bool operator!=(const __accessor_iterator_base<_Dimensions> &_Other) const {
+    return !(*this == _Other);
+  }
+
+private:
+  void __increment() {
+    if (_MLinearCurrent >= _MLinearEnd)
+      return;
+
+    ++_MLinearCurrent;
+    if (_MCurrent[_Index0] < _MEnd[_Index0])
+      _MCurrent[_Index0]++;
+    if constexpr (_Dimensions > 1) {
+      if (_MCurrent[_Index0] == _MEnd[_Index0]) {
+        if (_MCurrent[_Index1] < _MEnd[_Index1]) {
+          _MCurrent[_Index1]++;
+          _MCurrent[_Index0] = _MBegin[_Index0];
+        }
+      }
+    }
+    if constexpr (_Dimensions > 2) {
+      if (_MCurrent[_Index1] == _MEnd[_Index1]) {
+        if (_MCurrent[_Index2] < _MEnd[_Index2]) {
+          _MCurrent[_Index2]++;
+          _MCurrent[_Index0] = _MBegin[_Index0];
+          _MCurrent[_Index1] = _MBegin[_Index1];
+        }
```

Comment:
Is this simply an unroll? If so, can we use/generalize https://github.com/intel/llvm/blob/sycl/sycl/include/sycl/accessor.hpp#L229-L242 ?

Reply:
I don't think that this is an unroll. Here each next `if` depends on the previous and I guess in most cases the second and the third branches won't be ever executed.

If necessary, it could be rewritten using `dim_loop`, but I think that it will significantly reduce readability

Reply:
Looks like double nested triangular loop to me (both are compile-time unroll). But I won't insist.

Reply:
This is not actual anymore since ba4557784b2781a6b750b3489e1d6cf850b4a615